hudi-spark-datasource/hudi-spark/src/test/java/org/apache/hudi/functional/TestIncrementalQueries.java

```diff
@@ -0,0 +1,207 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.hudi.functional;
+
+import org.apache.hudi.DataSourceReadOptions;
+import org.apache.hudi.DataSourceWriteOptions;
+import org.apache.hudi.SparkAdapterSupport$;
+import org.apache.hudi.common.config.HoodieMetadataConfig;
+import org.apache.hudi.common.table.HoodieTableConfig;
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.table.timeline.HoodieInstant;
+import org.apache.hudi.common.util.collection.Pair;
+import org.apache.hudi.config.HoodieWriteConfig;
+import org.apache.hudi.testutils.SparkClientFunctionalTestHarness;
+
+import org.apache.spark.sql.Column;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.RowFactory;
+import org.apache.spark.sql.SaveMode;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class TestIncrementalQueries extends SparkClientFunctionalTestHarness {
+  @ParameterizedTest
+  @CsvSource({
+      "6,COPY_ON_WRITE", "8,COPY_ON_WRITE", "9,COPY_ON_WRITE",
+      "6,MERGE_ON_READ", "8,MERGE_ON_READ", "9,MERGE_ON_READ"})
+  void testIncrementalQueryWithMultiCommitsInSameFile(String tableVersion, String tableType) {
+    String path = basePath();
+    Map<String, String> hudiOptions = createWriteOption(tableVersion, tableType);
+    StructType schema = createTableSchema();
+    // Write three commits.
+    writeData(hudiOptions, path, tableVersion, tableType, schema);
+    // Get the commits and their 'commit time' based on table version.
+    HoodieTableMetaClient metaClient =
+        HoodieTableMetaClient.builder().setBasePath(path).setConf(storageConf()).build();
+    List<Pair<HoodieInstant, String>> sortedInstants = metaClient.getActiveTimeline().getInstants()
+        .stream()
+        .map(instant -> Pair.of(
+            instant, tableVersion.equals("6") ? instant.requestedTime() : instant.getCompletionTime()))
+        .collect(Collectors.toList());
+
+    // Run incremental query: last two commits are within the range.
+    // Make sure the records from the second commit are included.
+    // This avoids the differences between different versions. That is,
+    // the range type of table version 6 is open_close, but that of > 6 is close_close by default.
```

Comment:
hey @yihua : this open_close and close_close is an intentional design?
how does a user coming from older versions might know this?